### ISSUE TITLE
bedtools sort sometimes out of order

### ISSUE THREAD

**Issue by rvalieris** [arq5x/bedtools2#121]
Sometimes the output of bedtools sort will have records out of order.

I think this is related to repeated records on the bed, but I am not sure.

test case:

```
$ cat test.bed
chr1    27100144        27100235
chr1    27100174        27100182
chr1    27100204        27100206
chr1    27100205        27100205
chr1    27100204        27100206
chr1    27100213        27100215
chr1    27100275        27100397
chr1    27100276        27100397

$ bedtools --version
bedtools v2.21.0

$ bedtools sort -i test.bed > test.sorted.bed

$ sort -c -k1,1 -k2,2n test.sorted.bed
sort: test.sorted.bed:5: disorder: chr1   27100204        27100206
```

Furthermore, this affects other tools that need the bed to be sorted.

```
$ bedtools merge -i test.sorted.bed
Error: Sorted input specified, but the file test.sorted.bed has the following out of order record
chr1    27100204        27100206
```

**Comment by duartemolha** [comment]
I believe I have seen this behavior before... but since I only updated to a more recent version today I cannot be sure when this has occured... but I have had occasions that my input file for mergBed had just been sorted by sortBed and I was getting that very same error message

**Comment by LukeGoodsell** [comment]
This appears to be due to the zero-length interval (chr1:27100205-27100205) being mis-sorted. As I recently [reported in the bedtools-discuss mailinglist](https://groups.google.com/forum/#!topic/bedtools-discuss/vpHGXtQSFfA), zero-length intervals are not properly sorted. However, if you sort the input (with GNU coreutils' sort, for example) before running the file through sortBed, the output will be correctly ordered and will have undergone the validity-testing that sortBed applies.

Example:

    $ bedtools --version
    bedtools v2.25.0-3-gd2d9c95
    $ cat testin.bed
    chr1    30      40      RegionD 0       +
    chr1    25      25      RegionC 0       +
    chr1    10      20      RegionA 0       +
    chr1    24      30      RegionB 0       +
    $ sortBed -i "testin.bed" | tee test1.bed | sort -c -k1,1 -k2,2n || true
    sort: -:3: disorder: chr1       24      30      RegionB 0       +
    $ cat test1.bed
    chr1    10      20      RegionA 0       +
    chr1    25      25      RegionC 0       +
    chr1    24      30      RegionB 0       +
    chr1    30      40      RegionD 0       +
    $ sort -V -k 1,3 "testin.bed" | sortBed | tee test2.bed | sort -c -k1,1 -k2,2n || true
    $ cat test2.bed
    chr1    10      20      RegionA 0       +
    chr1    24      30      RegionB 0       +
    chr1    25      25      RegionC 0       +
    chr1    30      40      RegionD 0       +

**Comment by LukeGoodsell** [comment]
Here's the culprit: https://github.com/arq5x/bedtools2/blob/a0dc5db713af6d0bad420c8bb0dbd618e1bdb72e/src/utils/bedFile/bedFile.h#L691

Removing the `bed.start--;` line fixes this issue, but I don't know if it's needed for something. I see that it was added [here](https://github.com/arq5x/bedtools2/commit/4a64280e9a6b4b35ba825956b63b1beecd2a5b7a) to support zero-length intervals, but I don't see the rationale. It appears that the start and end co-ordinates are restored elsewhere under various circumstances (respecting the bed.zeroLength flag), so maybe the sortByStart function in bedField.cpp should respect it also?

@arq5x, I've made [a pull request that causes sortByStart to respect the zeroLength flag](https://github.com/arq5x/bedtools2/pull/409/files) and resolves the issue describe here.
